### PR TITLE
Refactor Dockerfile

### PR DESCRIPTION
**Pull request by ErikDeSmedt**
This commit improves the dev-container
- faster build-time
- improved readability

### CODE REVIEW

File: docker/gl-testing/Dockerfile [hunk 1]
```diff
@@ -1,219 +1,273 @@
...
+
+# Ensure we can use cargo under a different user
+RUN chmod a+rwx -R $CARGO_HOME
+
+
+# ------------------------------------------
+# STAGE: cln-downloader
+# -----------------------------------------
+# gl-testing requires multiple versions of Core Lightning
+# DOwnload all versions so they can be copied to our dev-container
```
Comment (Randy808) [inline docker/gl-testing/Dockerfile:111]: nit: `Download`

File: docker/gl-testing/Dockerfile [hunk 2]
```diff
@@ -1,219 +1,273 @@
...
-ENV DEVELOPER=1
-
-# Add the latest CLN version to the `$PATH` so `pyln-testing` can find
-# it when it wants to spawn a new node
-ENV PATH=/opt/cln-latest/usr/local/bin/:$PATH
-
-# And we also add the cargo bin directory just for the sake of
-# simplifying a dev's life.
-ENV PATH=/home/$DOCKER_USER/.cargo/bin:$PATH
+# COnfigure rust and related logging
```
Comment (Randy808) [inline docker/gl-testing/Dockerfile:219]: nit: `Configure`

File: docker/gl-testing/Dockerfile [hunk 3]
```diff
@@ -1,219 +1,273 @@
...
+ADD libs/gl-client-py/pyproject.toml /repo/libs/gl-client-py/pyproject.toml
+ADD libs/gl-testing/pyproject.toml /repo/libs/gl-testing/pyproject.toml
+
+WORKDIR repo
+
+# We don't want to reinstall depenendencies every time
+# a single line of library code changes.
+#
+# Poetry install requires us to ADD the library-code to
+# the dockerfil. So this approach doesn't work
```
Comment (Randy808) [inline docker/gl-testing/Dockerfile:68]: nit: `dockerfile`

**Comment by ErikDeSmedt** [comment]
@cdecker : The current failure is a flake. Could you re-run the pipeline

**Comment by cdecker** [comment]
Sure, making you an owner of the repo too 👍

**Review by Randy808 — APPROVED** [review]

File: docker/gl-testing/Dockerfile [hunk 4]
```diff
@@ -1,219 +1,273 @@
...
-    curl \
-    wget \
-    unzip \
-    && rm -rf /var/lib/apt/lists/*
 
+
+# --------------------------------------------------
+# STAGE: python-builder
+#--------------------------------------------------
+# This imaged is used to build the python environment
```
Comment (cdecker) [inline docker/gl-testing/Dockerfile:33]: `imaged` -> `image`

File: .github/workflows/check-self.yml [hunk 5]
```diff
@@ -31,8 +31,8 @@ jobs:
         key: ${{ runner.os }}-cargo-${{ hashFiles('**/Cargo.lock') }}
     - name: Check Self
       run : |
-        make docker-check-self
+        PYTEST_OPTS='-n 8' make docker-check-self
```
Comment (cdecker) [inline .github/workflows/check-self.yml:34]: I'm surprised that the GHA runners support this many tasks in parallel, good find :+1:

File: .github/workflows/check-self.yml [hunk 6]
```diff
@@ -22,6 +22,13 @@ jobs:
       run: |
         sudo chmod a+rwx -R libs/gl-client-py
         make docker-image
+    - name: Rust Cache
+      uses: actions/cache@v2
+      with:
+        path: |
+          /tmp/gltesting/cargo
+          /tmp/gltesting/target
+        key: ${{ runner.os }}-cargo-${{ hashFiles('**/Cargo.lock') }}
```
Comment (cdecker) [inline .github/workflows/check-self.yml:31]: I never quite understood how much cache is available and what the speed to access it is. Did you find details on this (retention, size, and cache speed)? Just out of curiosity.

Comment (ErikDeSmedt) [reply .github/workflows/check-self.yml:31]: I have no idea.

"If it works it works, it it doesn't it doesn't matter" is my current approach to Caching.

Comment (cdecker) [reply .github/workflows/check-self.yml:31]: Gotcha :+1:

**Comment by cdecker** [comment]
ACK https://github.com/Blockstream/greenlight/pull/390/commits/6b808e6faf3b3c3c5e3285d0bcdacd013a90c065

**Pull request merged**